botocore/endpoints_v2.py

```diff
@@ -0,0 +1,1123 @@
+            {
+            "fn": "isValidHostLabel",
+            "argv": [
+                {
+                "ref": "parameterName"
+                }
+            ],
+            "type": "tree",
+            "rules": [
+                {
+                "type": "tree",
+                "conditions": [
+                    // Abbreviated for clarity
+                ],
+                "rules": [
+                    // Abbreviated for clarity
+                ]
+                },
+                {
+                "type": "endpoint",
+                "conditions": [
+                    // Abbreviated for clarity
+                ],
+                "endpoint": {
+                    "url": "{parameterName}.amazonaws.com"
+                }
+                },
+                {
+                "type": "error",
+                "conditions": [
+                    // Abbreviated for clarity
+                ],
+                "error": "an error message"
+                }
+            ]
+            }
+        ]
+    }
+
+    A tree rule is non-terminal meaning it will never be returned to a provider.
+    Additionally this means it has no attributes that need to be resolved.
+    """
+
+    def __init__(self, rules, **kwargs):
+        super().__init__(**kwargs)
+        self.rules = [
+            RuleCreator.create(partitions=self.partitions, **rule)
+            for rule in rules
+        ]
+
+    def evaluate(self, scope_vars):
+        """If a tree rule's conditions evaluate successfully, iterate over its
+        subordinate rules and return a result if there is one. If any of the
+        subsequent rules are trees, the function will recurse until it reaches
+        an error or an endpoint rule.
+
+        :type scope_vars: dict
+        :param scope_vars: A dictionary combining input parameters and variables
+        that have been scoped throughout the rule's evaluation.
+
+        :returns: An endpoint or error object.
+        """
+        if self.evaluate_conditions(scope_vars) is True:
+            for rule in self.rules:
+                # newly set parameters via "assign" cannot be shared between
+                # adjacent rules
+                evaluation = rule.evaluate(scope_vars.copy())
+                if evaluation:
+                    return evaluation
+        return None
+
+
+class RuleCreator:
+    """A helper that creates rule objects."""
+
+    endpoint = EndpointRule
+    error = ErrorRule
+    tree = TreeRule
+
+    @classmethod
+    def create(cls, **kwargs):
+        """Given a rule dictionary, locate the `type` attribute, using it to
+        determine the type of rule object to create. Then create and return it.
+
+        :type kwargs: dict
+        :params kwargs: An arbitrary dictionary containing required parameters
+        for different types of rule objects such asconditions, rules, errors,
+        endpoints etc.
+
+        :raises: RuleSetDefinitionError if the type is not recognized.
```

Comment:
This is also probably redundant since it's apparent in the code and we're not generating docs with this.